### PR TITLE
Counting CXL errors for Correctable, Uncorrectable, and CFG err type

### PR DESCRIPTION
**Pull request by gadhiadevarsh**
Summary:  VISTARTA: Counting CXL errors for Correctable, Uncorrectable, and CFG err type

Test Plan: 1) Inject errors by cxl pci-err-inj
                2) get the logs on Uart terminal to confirm if the error is injected
                3) uart shell command to get the count update [ cxl list_errors]
               4) execute implemented cxl command cxl-err-cnt-get to get the same out put on cxl terminal

Reviewers:

Subscribers:

Tasks: T191983816 : implement cxl error counters on device side

Tags:

### CODE REVIEW

File: cxl/lib/libcxl.c [hunk 1]
```diff
@@ -14542,6 +14542,198 @@ CXL_EXPORT int cxl_memdev_ddr_page_select_get(struct cxl_memdev *memdev)
...
+  REINIT_THRESHOLD,
+  ENCOD_VIOLATION,
+  POISON_RVD,
+  RCVR_OVERFLOW,
+  RSVDP_12,
+  RSVDP_13,
+  CXL_INTERNAL_ERR,
+  CXL_IDE_TX_ER,
+  CXL_IDE_RX_ER,
+  UNCORR_ERR_COUNT = 17
```
Comment (lnishal) [inline cxl/lib/libcxl.c]: By default this value will be 17, any specific reason to mention the number? Same comment on the naming of this enum. Same comment for other enums as well

File: cxl/lib/libcxl.c [hunk 2]
```diff
@@ -14542,6 +14542,198 @@ CXL_EXPORT int cxl_memdev_ddr_page_select_get(struct cxl_memdev *memdev)
...
+// CXL_TOP_CXL_CONT_GENERIC_INTERRUPT_STATUS_04[12:6]
+enum cxl_corr_errors {
+  CORR_CACHE_DATA_ECC = 0,
+  CORR_MEM_DATA_ECC,
+  CORR_CRC_THRESHOLD,
+  CORR_RETRY_THRESHOLD,
+  CACHE_POISON_RVD,
+  MEM_POISON_RVD,
+  PHY_LAYER_ERR,
+  CORR_ERR_COUNT = 7
```
Comment (lnishal) [inline cxl/lib/libcxl.c]: By default this value will be 7, any specific reason to mention the number? Can we rename this to indicate max error types say `MAX_CORR_ERR_COUNT` ?

File: cxl/lib/libcxl.c [hunk 3]
```diff
@@ -14542,6 +14542,198 @@ CXL_EXPORT int cxl_memdev_ddr_page_select_get(struct cxl_memdev *memdev)
     return rc;
 }
 
+/* CXL ERROR COUNT GET */
+#define CXL_MEM_COMMAND_ID_CXL_ERROR_CNT_GET CXL_MEM_COMMAND_ID_RAW
+#define CXL_MEM_COMMAND_ID_CXL_ERROR_CNT_GET_OPCODE 0xFB23
```
Comment (Balaji-kalidass) [inline cxl/lib/libcxl.c]: it should be same as opcode you have used in the firmware side.

File: cxl/lib/libcxl.c [hunk 4]
```diff
@@ -14542,6 +14542,198 @@ CXL_EXPORT int cxl_memdev_ddr_page_select_get(struct cxl_memdev *memdev)
...
+    "DL_PROTOCOL_ERR_STS",
+    "CORRECTED_INTERNAL_ERR_STS",
+    "BAD_TLP_ERR_STS",
+    "BAD_DLLP_ERR_STS"};
+
+
+struct cxl_mbox_handle_cxl_err_cnt_out {
+  uint32_t total_err_cnt;
+  uint8_t corr_err[CORR_ERR_COUNT];
+  uint8_t resvd1;
```
Comment (Balaji-kalidass) [inline cxl/lib/libcxl.c]: why you have used the reserved here?

File: cxl/lib/libcxl.c [hunk 5]
```diff
@@ -14542,6 +14542,198 @@ CXL_EXPORT int cxl_memdev_ddr_page_select_get(struct cxl_memdev *memdev)
...
+        fprintf(stderr, "%s: cxl_cmd_new_raw returned Null output\n",
+                cxl_memdev_get_devname(memdev));
+        return -ENOMEM;
+    }
+
+    query = cmd->query_cmd;
+    cinfo = &query->commands[cmd->query_idx];
+
+    /* used to force correct payload size */
+    cinfo->size_in = CXL_MEM_COMMAND_ID_LOG_INFO_PAYLOAD_IN_SIZE;
```
Comment (lnishal) [inline cxl/lib/libcxl.c]: This value should be 0, as there are no input bytes for this command.

**Review by lnishal — CHANGES_REQUESTED** [review]

File: cxl/builtin.h [hunk 6]
```diff
@@ -18,8 +18,10 @@ int cmd_check_labels(int argc, const char **argv, struct cxl_ctx *ctx);
 int cmd_identify(int argc, const char **argv, struct cxl_ctx *ctx);
 int cmd_get_supported_logs(int argc, const char **argv, struct cxl_ctx *ctx);
 int cmd_get_log(int argc, const char **argv, struct cxl_ctx *ctx);
-int cmd_get_event_interrupt_policy(int argc, const char **argv, struct cxl_ctx *ctx);
-int cmd_set_event_interrupt_policy(int argc, const char **argv, struct cxl_ctx *ctx);
+int cmd_get_event_interrupt_policy(int argc, const char **argv,
```
Comment (lnishal) [inline cxl/builtin.h]: Did you run clang format on this file? I think there are many lines modified and showing in diff without any code change

**Review by lnishal — CHANGES_REQUESTED** [review]

File: cxl/memdev.c [hunk 7]
```diff
@@ -4373,12 +4378,13 @@ static int action_cmd_get_cxl_membridge_stats(struct cxl_memdev *memdev,
 {
 	if (cxl_memdev_is_active(memdev)) {
 		fprintf(stderr, "%s: memdev active, abort cxl membridge stats\n",
-    			cxl_memdev_get_devname(memdev));
+                cxl_memdev_get_devname(memdev));
```
Comment (gadhiadevarsh) [inline cxl/memdev.c]: git commit was giving space related error here line 4381 and 4236, and hence you would see space / tab related changes but those changes are not related to code

Comment (lnishal) [reply cxl/memdev.c]: not sure why these are giving errors.

File: cxl/memdev.c [hunk 8]
```diff
@@ -4373,12 +4378,13 @@ static int action_cmd_get_cxl_membridge_stats(struct cxl_memdev *memdev,
...
 		return -EBUSY;
 	}
   return cxl_memdev_get_cxl_membridge_stats(memdev);
 }
 
+
```
Comment (lnishal) [inline cxl/memdev.c]: remove unintended new line

File: cxl/lib/libcxl.c [hunk 9]
```diff
@@ -15127,3 +15127,202 @@ CXL_EXPORT int cxl_memdev_ddr_refresh_mode_get(struct cxl_memdev *memdev)
     cxl_cmd_unref(cmd);
     return rc;
 }
+/* CXL ERROR COUNT GET */
```
Comment (lnishal) [inline cxl/lib/libcxl.c:15131]: Please add a new line before this new block

**Comment by lnishal** [comment]
LGTM

**Comment by Balaji-kalidass** [comment]
LGTM

File: cxl/memdev.c (hunk 7 above)
Comment (lnishal) [reply cxl/memdev.c]: Looks some git pre-hook commits got added. after removing those it got resolved

**Comment by sandeepa-prabhu-meta** [comment]
LGTM
